### ISSUE TITLE
Tmux wait-for and signaling

### ISSUE THREAD

**Issue by bracco23** [tmux/tmux#832]
Hello, messing with tmux, scripting ad the wait-for functionality.

**LONG VERSION, TL;DR VERSION DOWN**

What I was trying to do was to use tmux to access a certain number of processes in a certain order. So I basically have a script that run a session and the, each time the script is invoked, adds a window to that session with the process I want to run. The script is run as a prefix inside another launching tool (roslaunch, part of the ROS framework).

Now, the issue is this:

- Roslaunch keeps track of each process it launches, so I use tmux signaling to keep the bash script process alive while the actual process inside tmux is alive (and die when the process dies.

- Roslaunch use Unix signals to communicate with processes, basically usually to stop processes I send a SIGINT and it is forwarded to every process run by Roslaunch, causing it to do stuff.

Now, since I don't directly run the process itself, but I have this intermediate layer that put in tmux, what I wanted to do is trap the SIGINT signal and forward inside tmux using **send-keys**.

But now the not working part: trap only execute its handles at the end of the actual command executed, and if this command is **tmux wait-for**, well the end is not coming that soon.

So, is there a way to have the **wait-for** function to stop on signals like SIGINT to have other stuff being executed? seems like a good idea to me.

Another thing that might work is to add a *timeout* option so that we can say:
    wait for a maximum of n second then exit and return 1 instead of zero.

The second way could allow me to like put the **wait-for** in loop and have the trap check between executions to allow me a limited delay for the signal to be forwarded.

**TL;DR VERSION**

I'm trying to mess with **wait-for** and **trap**, but **wait-for** waits until the signal comes and trap doesn't execute the handler for the Unix signal till the actual command is over, so basically if I'm waiting for a tmux signal I cannot receive a Unix signal. Can it be handled so that tmux stops waiting if a Unix signal is sended? Or can a timeout be added so that a workaround for this case can be created?

Info for my setup:

**uname -sp && tmux -V && echo $TERM**
Linux x86_64
tmux 1.8
xterm *outside*
screen *inside*

**Comment by nicm** [comment]
No there is no way to do this currently. Timeout would probably be
easier to implement than signals. I suggest adding a -T flag to
wait-for, you will need to store the timeout in the cmdq_item and check
it when CMDQ_WAITING is set in cmdq_next, as well as adding a libevent
timer to ensure the timeout is checked often enough.

But you will need to work on Git master, 1.8 is far too old.

**Comment by bracco23** [comment]
I don't want to sound the guy that tells what's wrong and leaves, but I'm currently far too busy to implement it myself, I'm working around it and for now it is fine for me, I just thought that this might be somehow a useful feature.

**Comment by nicm** [comment]
OK well I don't need this particularly so I don't think I will be working on it. I'll put it in the TODO file and perhaps someone will look at it.
